### ISSUE TITLE
Waypoint config does not always restart in Nomad container deployment 

### ISSUE THREAD

**Issue by stevenheinze57** [hashicorp/waypoint#2721]
I've noticed with "waypoint config sync" when running this against a nomad deployment that uses the "nomad-jobspec" plugin it will not always restart the container, actually more often than not, it does not restart but instead updates the applications env in place. This is an issues depending on how my application is handling those environment configurations (sometimes a restart is necessary). Is this a known issue with this plugin? 

Waypoint version: 0.6.2

**Comment by briancain** [comment]
Hey there @stevenheinze57 - Could you please share an example that demonstrates the behavior you mentioned? Additionally, please share Waypoint Server + Runner logs and Nomad logs too.

I used this example project: https://github.com/hashicorp/waypoint-examples/tree/main/nomad/nodejs-jobspec
But updated the config to have a new ENV var:

```hcl
project = "example-nodejs"

config {
  env = {
    WP_TEST = "test-string"
  }
}

...
...
```

In my own testing, using latest Nomad + Waypoint, config sync updates the containers environment var, and then eventually a few seconds later restarts the container. This seems to match the behavior shown in the docs: https://tip.waypointproject.io/docs/app-config#application-restart-behavior

It looks like Waypoint does restart the container, just not immediately once `waypoint config sync` finishes up. It's a few seconds later. Is this the behavior you are seeing?

**Comment by stevenheinze57** [comment]
@briancain   Looks like I was able to find the behavior in the entrypoint logs. What I was expecting was it to be demonstrated in the nomad UI that the container restart took place, instead this seems to update the process within the container as opposed to the entire container itself, or calling upon nomad to orchestrate the restart. This seems like it would nullify any restart behavior that may be configured on the nomad jobspec. 

2021-12-03T19:29:53.790Z [DEBUG] entrypoint.config_recv: new configuration received
2021-12-03T19:29:53.791Z [DEBUG] entrypoint.config.watcher.watchloop: new config variables received, scheduling refresh
2021-12-03T19:29:54.292Z [DEBUG] entrypoint.config.watcher.watchloop: new configuration computed
2021-12-03T19:29:54.292Z [INFO]  entrypoint.config.watcher: env vars changed, sending new child command
2021-12-03T19:29:54.292Z [DEBUG] entrypoint.child: child command received
2021-12-03T19:29:54.292Z [INFO]  entrypoint.child: terminating current child process for restart
2021-12-03T19:29:54.292Z [DEBUG] entrypoint.child: sending SIGTERM: pid=15
2021-12-03T19:29:54.293Z [DEBUG] entrypoint.child: SIGTERM sent, waiting for child process to end or timeout: pid=15
2021-12-03T19:29:54.303Z [WARN]  entrypoint.child: subprocess exited: args=["/usr/local/bin/docker-entrypoint.sh", "npm", "run", "debug-server"] cmd=/usr/local/bin/docker-entrypoint.sh err="signal: terminated"
2021-12-03T19:29:54.303Z [INFO]  entrypoint.child: child process exited: pid=15 wait_err=<nil>
2021-12-03T19:29:54.303Z [INFO]  entrypoint.child: starting child process: args=["/usr/local/bin/docker-entrypoint.sh", "npm", "run", "debug-server"] cmd=/usr/local/bin/docker-entrypoint.sh

**Comment by izaaklauer** [comment]
@stevenheinze57 You're right - when config is updated, waypoint doesn't call back to the nomad layer to trigger the container restart. Instead, the custom entrypoint binary (which is the container's init process) SIGTERMs the application process and then starts it anew when it notices the config has changed (which is on a 15 second poll).

From the app's perspective, this should be identical to a nomad level restart. If you need to wait significantly more than 15 seconds for the app to restart (or if the restart never happens), let us know and we can dig further into this issue.

**Comment by stevenheinze57** [comment]
@izaaklauer  I think an ideal option to have without getting too granular at this point would be a parameter set on the waypoint.hcl file to allow restarts to be orchestrated by the deploy driver (in this case "nomad-jobspec"). There are a number of configurations for restarts with nomad deployments that I think are very beneficial that don't seem present within waypoint yet. 

One thing I can think of immediately is staggered updates. From what I'm understanding here, waypoint is doing a config update across all these container simultaneously, whereas this is a restart behavior already able to be defined by nomad.
